### PR TITLE
[XPU] ep+tp all2all

### PR DESCRIPTION
**Pull request by zhupengyang**
## Motivation

- FD_EP_TP_STRATEGY="all_reduce" or "all_to_all"
image

## Usage or Command

```
export FD_EP_TP_STRATEGY="all_to_all"
python -m fastdeploy.entrypoints.openai.api_server \
  --model <model_path> \
  --port 8188 \
  --tensor-parallel-size 8 \
  --enable-expert-parallel \
  --data-parallel-size 8 \
  --max-model-len 32768 \
  --max-num-seqs 64 \
  --quantization "wint4" \
  --engine-worker-queue-port "8023,8033,8043,8053,8063,8073,8083,8093" \
  --gpu-memory-utilization 0.9 \
  --load-choices "default"
```

## Checklist

- [ ] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [ ] Format your code, run `pre-commit` before commit.
- [ ] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/config.py [hunk 1]
```diff
@@ -575,6 +575,11 @@ def __init__(
         else:
             self.pd_disaggregation_mode = "None"
 
+        # ep+tp split strategy
+        # 0: qkv_linear + attn + out_linear + allreduce
+        # 1: allgather + qkv_linear + attn + all2all + out_linear
+        self.ep_tp_split_mode = int(os.getenv("EP_TP_SPLIT_MODE", 0))
```
Comment (yuanlehome) [inline fastdeploy/config.py]: 新增环境变量放到envs.py中，命名以FD_开头，并提供注释

File: fastdeploy/worker/xpu_model_runner.py [hunk 2]
```diff
@@ -102,6 +103,7 @@ def xpu_pre_process(
         block_tables=share_inputs["block_tables"],
         caches=share_inputs["caches"],
     )
+    xpu_forward_meta.token_num = token_num
```
Comment (yuanlehome) [inline fastdeploy/worker/xpu_model_runner.py]: forward_meta中添加字段需要在data class增加，就这一行而言，不建议写，直接通过xpu_forward_meta.ids_remove_padding.shape[0]就是token_num

File: fastdeploy/config.py (hunk 1 above)
Comment (yuanlehome) [reply fastdeploy/config.py]: 我也建议是字符串类型，值可以是 all_reduce 或 all_to_all

File: fastdeploy/model_executor/load_weight_utils.py [hunk 3]
```diff
@@ -267,8 +268,14 @@ def get_expert_ranges(fd_config):
...
     if fd_config.parallel_config.tensor_parallel_size > 1:
+        no_tp_action_keys = copy.deepcopy(num_local_ffn_keys)
+        if fd_config.parallel_config.ep_tp_split_mode == 1:
+            for i in range(fd_config.model_config.moe_layer_start_index, fd_config.model_config.num_hidden_layers):
+                k = f"ernie.layers.{i}.self_attn.o_proj.weight"
+                if k in weight_list:
+                    no_tp_action_keys.append(k)
         tp_actions = cls._get_tensor_parallel_mappings(fd_config.model_config.pretrained_config)
-        new_actions = {k: v for k, v in tp_actions.items() if k not in num_local_ffn_keys}
+        new_actions = {k: v for k, v in tp_actions.items() if k not in no_tp_action_keys}
```
Comment (yuanlehome) [inline fastdeploy/model_executor/load_weight_utils.py:278]: 这里记上一个TODO，V1 loader逻辑是不会走到这里的，需要适配V1 loader

File: fastdeploy/model_executor/layers/normalization.py [hunk 4]
```diff
@@ -165,10 +235,18 @@ def forward(self, x, residual_input: Optional[paddle.Tensor] = None) -> paddle.T
...
+        if self.split_x:
+            residual_out = self.split(residual_out)
+        if self.allgather_out:
+            out = self.allgather(out, forward_meta.token_num)
+
+        if residual_input is None:
+            return out
         else:
-            return norm_out[0].astype(x_dtype)
+            return out, residual_out
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/normalization.py:248]: 这个是不是可以写到linear层里去，放到这里太奇怪了，norm层感知到tp/ep，不合理

File: fastdeploy/model_executor/layers/linear.py [hunk 5]
```diff
@@ -815,14 +816,29 @@ def __init__(
...
+            self.ep_size > 1
+            and self.tp_size > 1
+            and fd_config.parallel_config.ep_tp_split_mode == 1
+            and layer_id >= fd_config.model_config.moe_layer_start_index
+            and layer_id < fd_config.model_config.num_hidden_layers
+        )
+        if self.split_token:
+            self.num_heads = fd_config.model_config.num_attention_heads
+        else:
+            self.num_heads = fd_config.model_config.num_attention_heads // self.nranks
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/linear.py]: num_heads这块是不是可以删掉？row parallel linear不需要这个变量？后面需要self.num_heads * self.head_dim的话，是不是self.hidden_size就可以 ？

File: fastdeploy/model_executor/layers/linear.py [hunk 6]
```diff
@@ -854,13 +870,30 @@ def __init__(
 
         self.reduce_results = reduce_results
 
+    def all2all_transpose(self, x: paddle.Tensor) -> paddle.Tensor:
+        token_num = x.shape[0]
+        token_num_pad = (token_num + self.tp_size - 1) // self.tp_size * self.tp_size
+        if token_num_pad > token_num:
+            x_new = paddle.zeros([token_num_pad, x.shape[1]], x.dtype)
+            x_new[:token_num, :] = x
+            x = x_new
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/linear.py:875]: 这一块不用if判断了吧，否则cudagraph捕获不了

File: fastdeploy/model_executor/layers/linear.py [hunk 7]
```diff
@@ -799,6 +799,7 @@ def __init__(
         reduce_results: bool = True,
         skip_quant: bool = False,
         weight_dtype="",
+        layer_id: int = -1,
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/linear.py:802]: 这个新增参数，建议所有model.py里都传一下吧，默认为-1不合理

File: fastdeploy/model_executor/layers/normalization.py [hunk 8]
```diff
@@ -124,7 +150,51 @@ def load_state_dict(self, state_dict: Dict[str, paddle.Tensor | np.ndarray]):
...
+        Args:
+            out (paddle.Tensor): Output tensor to be gathered.
+
+        Returns:
+            paddle.Tensor: Gathered tensor.
+        """
+        token_num_per_rank = out.shape[0]
+        multi_outs = paddle.zeros([token_num_per_rank * self.tp_size, out.shape[1]], dtype=out.dtype)
+        paddle.distributed.all_gather(multi_outs, out, self.tp_group)
+        out = multi_outs if token_num is None else multi_outs[:token_num, :]
```
Comment (yuanlehome) [inline fastdeploy/model_executor/layers/normalization.py]: 这里也会导致cudagraph捕获不了，默认就multi_outs[:token_num, :] ？

File: fastdeploy/model_executor/layers/normalization.py (hunk 4 above)
Comment (zhupengyang) [reply fastdeploy/model_executor/layers/normalization.py:248]: 这里需要切分residual_out，只能在norm中感知

File: fastdeploy/model_executor/layers/linear.py (hunk 7 above)
Comment (zhupengyang) [reply fastdeploy/model_executor/layers/linear.py:802]: 当前方案需要在layer层内部确定在组网的什么位置。

所有的out_linear都是用的RowParallelLinear，通过传入layer_id确定位置，这种方式比较trick。

比如mlp和shared_expert都是用的RowParallelLinear，都传入layer_id的话，就还需要其他信息来确定当前layer是不是out_linear

可能得看下怎么更准确得描述layer在模型的哪个位置，比如传入name string来感知当前位置

File: fastdeploy/model_executor/load_weight_utils.py (hunk 3 above)
Comment (zhupengyang) [reply fastdeploy/model_executor/load_weight_utils.py:278]: 适配V1 loader的时候会再统一处理

File: fastdeploy/worker/xpu_model_runner.py (hunk 2 above)
Comment (zhupengyang) [reply fastdeploy/worker/xpu_model_runner.py]: done

File: fastdeploy/model_executor/layers/linear.py (hunk 5 above)
Comment (zhupengyang) [reply fastdeploy/model_executor/layers/linear.py]: done

File: fastdeploy/model_executor/layers/normalization.py (hunk 8 above)
Comment (zhupengyang) [reply fastdeploy/model_executor/layers/normalization.py]: done

File: fastdeploy/config.py (hunk 1 above)
Comment (zhupengyang) [reply fastdeploy/config.py]: done

**Review by hong19860320 — APPROVED** [review]
LGTM

**Review by yuanlehome — APPROVED** [review]

**Comment by yuanlehome** [comment]
PR描述里给一下启动脚本示例吧～

File: fastdeploy/model_executor/layers/normalization.py [hunk 9]
```diff
@@ -165,10 +234,18 @@ def forward(self, x, residual_input: Optional[paddle.Tensor] = None) -> paddle.T
...
             )
-        if residual_input is not None:
-            return norm_out[0].astype(x_dtype), norm_out[1].astype(residual_input_dtype)
+        out = norm_out[0].astype(x_dtype)
+        residual_out = norm_out[1].astype(residual_input_dtype) if residual_input is not None else None
+
+        if self.split_x:
+            residual_out = self.split(residual_out)
+        if self.allgather_out:
+            out = self.allgather(out, forward_meta.ids_remove_padding.shape[0])
```
Comment (YuanRisheng) [inline fastdeploy/model_executor/layers/normalization.py:243]: 这俩分支场景现在有单测能测到么

Comment (zhupengyang) [reply fastdeploy/model_executor/layers/normalization.py:243]: 有xpu的单测。gpu目前跑不到这里，需要gpu同学自己适配并验证

**Review by EmmonsCurse — APPROVED** [review]
